File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/StackTaskProvider.java
package com.example.alexeyglushkov.taskmanager.task;

import android.os.Handler;

import java.util.List;

/**
 * Created by alexeyglushkov on 13.08.16.
 */
public class StackTaskProvider extends SimpleTaskPool implements TaskProvider, Task.StatusListener {
    private String providerId;
    private int priority;

    private boolean areTasksDependent; //if enabled the top task blocks the next task until the former finishes
    private boolean isBlocked;

    public StackTaskProvider(boolean areTasksDependent, Handler handler, String id) {
        super(handler);

        this.providerId = id;
        this.areTasksDependent = areTasksDependent;
    }

    @Override
    public void setTaskProviderId(String id) {
        providerId = id;
    }

    @Override
    public String getTaskProviderId() {
        return providerId;
    }

    @Override
    public void setPriority(int priority) {
        this.priority = priority;
    }

    @Override
    public int getPriority() {
        return priority;
    }

    @Override
    protected void triggerOnTaskAdded(Task task) {
        if (!isBlocked) {
            super.triggerOnTaskAdded(task);
        }
    }

    @Override
    public Task getTopTask(List<Integer> typesToFilter) {
        Task result = null;
        if (canTakeTask()) {
            int index = getTopTaskIndex(typesToFilter);
            if (index != -1) {
                result = getTasks().get(index);
            }
        }

        return result;
    }

    @Override
    public Task takeTopTask(List<Integer> typesToFilter) {
        Task result = null;
        if (canTakeTask()) {
            int index = getTopTaskIndex(typesToFilter);
            result = getTasks().get(index);
            getTasks().remove(index);
            onTaskTaken(result);
        }

        return result;
    }

    private void onTaskTaken(Task task) {
        if (areTasksDependent) {
            isBlocked = true;

            task.addTaskStatusListener(new Task.StatusListener() {
                @Override
                public void onTaskStatusChanged(Task task, Task.Status oldStatus, Task.Status newStatus) {
                    if (Tasks.isTaskCompleted(task)) {
                        isBlocked = false;

                        if (getTaskCount() > 0) {
                            // TODO: trigger for the next tasks until not blocked
                            triggerOnTaskAdded(getTasks().get(0));
                        }
                    }
                }
            });
        }

        triggerOnTaskRemoved(task);
    }

    private void triggerOnTaskRemoved(Task task) {
        checkHandlerThread();

        for (TaskPoolListener listener : listeners) {
            listener.onTaskRemoved(StackTaskProvider.this, task);
        }
    }

    private boolean canTakeTask() {
        return !areTasksDependent || !isBlocked;
    }

    private int getTopTaskIndex(List<Integer> typesToFilter) {
        int result = -1;
        for (int i=0; i<getTasks().size(); ++i) {
            Task task = getTasks().get(i);

            boolean passFilter = typesToFilter == null || !typesToFilter.contains(task.getTaskType());
            if (!task.isBlocked() && passFilter) {
                result = i;
                break;
            }
        }

        return result;
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/StackTaskProvider.kt
package com.example.alexeyglushkov.taskmanager.task

import androidx.annotation.WorkerThread

/**
 * Created by alexeyglushkov on 13.08.16.
 */
open class StackTaskProvider(private val areTasksDependent: Boolean, //if enabled the top task blocks the next task until the former finishes
                             override var taskProviderId: String,
                             threadRunner: ThreadRunner): ListTaskPool(threadRunner), TaskProvider, Task.StatusListener {
    override var priority: Int = 0
    private var isBlocked: Boolean = false
    override var taskFilter: TaskProvider.TaskFilter? = null

    override fun triggerOnTaskAdded(task: Task) {
        if (!isBlocked) {
            super.triggerOnTaskAdded(task)
        }
    }

    @WorkerThread
    override fun getTopTask(): Task? {
        var result: Task? = null
        if (canTakeTask()) {
            val index = getTopTaskIndex()
            if (index != -1) {
                result = _tasks[index]
            }
        }

        return result
    }

    @WorkerThread
    override fun takeTopTask(): Task? {
        var result: Task? = null
        if (canTakeTask()) {
            val index = getTopTaskIndex()
            result = _tasks[index]
            _tasks.removeAt(index)
            onTaskTaken(result)
        }

        return result
    }

    @WorkerThread
    private fun onTaskTaken(task: Task) {
        if (areTasksDependent) {
            isBlocked = true

            task.addTaskStatusListener(object : Task.StatusListener {
                override fun onTaskStatusChanged(task: Task, oldStatus: Task.Status, newStatus: Task.Status) {
                    if (Tasks.isTaskFinished(task)) {
                        isBlocked = false

                        if (getTaskCount() > 0) {
                            // TODO: trigger for the next tasks until not blocked
                            triggerOnTaskAdded(_tasks[0])
                        }
                    }
                }
            })
        }

        onTaskRemoved(task)
    }

    private fun canTakeTask(): Boolean {
        return !areTasksDependent || !isBlocked
    }

    private fun getTopTaskIndex(): Int {
        var result = -1
        for (i in 0 until _tasks.size) {
            val task = _tasks[i]
            val taskFilter = taskFilter
            val isFiltered = taskFilter != null && taskFilter.isFiltered(task)

            if (!task.isBlocked() && !isFiltered) {
                result = i
                break
            }
        }

        return result
    }

    override fun tag() = "StackTaskProvider"
}
